PythonAPI/examples/autoware_demo.py

```diff
@@ -306,67 +479,78 @@ def main():
-    settings.substepping = False
-    world.apply_settings(settings)
-
-    # Autoware will be publishing TF information based on the URDF files
-    # of the vehicle and sensor kit. Disable TF publishing in CARLA
-    # to avoid conflicts.
-    world.set_publish_tf(False)
-
+    # Determine TimeStep Data to be used
+    time_step_info = TimeStepData(synchronous_mode=(not args.run_async),
+                                  hz_rate=args.hz_rate,
+                                  phys_substepping=args.substepping)
+
+    # Apply Settings
+    apply_world_settings(client=client,
+                         world=world,
+                         time_step_info=time_step_info,
+                         map_name=args.load_map,
+                         force_map_reload=args.force_reload)
+    log_info(
+        f"Applied settings:\n"
+        f"\tsynchronous mode: {time_step_info.synchronous_mode}\n"
+        f"\tfixed time step: {time_step_info.hz_rate} (in Hz)\n"
+        f"\tsimulation dt: {time_step_info.get_sim_dt()} (in seconds)\n"
+        f"\tphysics substepping: {time_step_info.phys_substepping}\n"
+        f"\ttime scale: {args.time_scale:.2f}\n"
+        f"\tpure step execution: {time_step_info.is_pure_step_execution_enabled()}"
+    )
+
+    # Spawn Ego
     spawn_point = random.choice(world.get_map().get_spawn_points())
     ego = spawn_ego_with_sensors(world, spawn_point)
-    info('Ego spawned!')
-    
-    world.tick() # tick to process the changes (settings, ego + sensors spawn)
-    move_spectator(world, ego)
-
-    info("Simulation time scale is %f" % args.time_scale)
-    warning('Kill this script before stopping simulation!')
-
-    frame_count = 0
-
-    start_time = time.time()
-    last_behind_error_time = start_time
-    # Tick world to follow desired time_scale
-    try:
-        while True:
-            current_time = time.time()
-            real_time_passed = current_time - start_time
-            sim_time_passed = real_time_passed * args.time_scale
-            desired_frame_count = math.floor(sim_time_passed * DESIRED_SIM_RATE)
 
-            if desired_frame_count <= frame_count:
-                continue
+    world.tick()  # tick to process the changes (settings, ego + sensors spawn)
+    move_spectator(world, ego)
 
-            frame_behind_count = max(desired_frame_count - frame_count, 0)
-            if (frame_behind_count > 10) and \
-                (current_time > last_behind_error_time + 2.0):  # Error every 2 seconds
-                    last_behind_error_time = current_time
-                    error(f'Simulation cannot keep up with the desired time scale!!! ({frame_behind_count} frames behind)')
+    log_info('Ego spawned!')
+    log_warning('Kill this script before stopping simulation!')
 
-            world.tick()
-            frame_count += 1
+    # Run simulation loop
+    time_step_info.synchronous_mode and run_sync_simulation_loop(target_sim_dt=time_step_info.get_sim_dt(),
```

Comment:
```suggestion
    if time_step_info.synchronous_mode: run_sync_simulation_loop(target_sim_dt=time_step_info.get_sim_dt(),
```

Even though these two are functionally the same, I think the "if" provides more semantic clarity when reading the source code.